File: oz/csrc/oos.cpp
#include "util.h"
#include "hash.h"

#include "oos.h"

#include <cassert>
#include <algorithm>
#include <iterator>
#include <set>

#include <boost/container/pmr/polymorphic_allocator.hpp>
#include <boost/container/pmr/monotonic_buffer_resource.hpp>

namespace oz {

using namespace std;

using boost::container::pmr::monotonic_buffer_resource;
using boost::container::pmr::polymorphic_allocator;

static action_prob_t sample_chance(const history_t &history, rng_t& rng,
                                   game_t::action_prob_allocator_t alloc);

void oos_t::search_t::prepare_suffix_probs() {
  Expects(suffix_prob_.x > 0);

  prob_t s1 = prefix_prob_.s1 * suffix_prob_.x;
  prob_t s2 = prefix_prob_.s2 * suffix_prob_.x;

  suffix_prob_.l = delta_*s1 + (1.0 - delta_)*s2;
  suffix_prob_.u = history_.utility(search_player_);

  Ensures(suffix_prob_.l > 0);
}

void oos_t::search_t::tree_step(action_prob_t ap) {
  const auto acting_player = history_.player();
  const auto alloc = get_allocator();

  const auto infoset = (acting_player != CHANCE) ?
                       history_.infoset(alloc) :
                       null_infoset();

  tree_step(ap, infoset);
}

void oos_t::search_t::tree_step(action_prob_t ap, const infoset_t &infoset) {
  Expects(state_ == state_t::SELECT || state_ == state_t::CREATE);
  Expects(!history_.is_terminal());

  // TODO does a zero pr_a make sense here?
  Expects(0 <= ap.pr_a && ap.pr_a <= 1);
  Expects(0 <  ap.rho1 && ap.rho1 <= 1);
  Expects(0 <  ap.rho2 && ap.rho2 <= 1);

  Expects(is_normal(ap.pr_a));
  Expects(is_normal(ap.rho1));
  Expects(is_normal(ap.rho1));

  const auto acting_player = history_.player();

  path_.emplace_back(path_item_t {
      acting_player, infoset,
      ap, prefix_prob_
  });

  // update prefix and sample probabilities
  if(acting_player == search_player_) {
    prefix_prob_.pi_i *= ap.pr_a;
  }
  else {
    prefix_prob_.pi_o *= ap.pr_a;
  }

  prefix_prob_.s1 *= ap.rho1;
  prefix_prob_.s2 *= ap.rho2;

  // move game forward one ply
  history_.act(ap.a);
}

auto oos_t::search_t::sample_tree(const tree_t &tree,
                                  const infoset_t &infoset,
                                  rng_t &rng) const
  -> tree_t::sample_ret_t
{
  const auto acting_player = history_.player();

  prob_t eps, gamma;
  if (acting_player == search_player_) {
    eps = eps_;
    gamma = 0;
  }
  else {
    eps = 0;
    gamma = gamma_;
  }

  const auto targets = (target_ && target_infoset_) ?
                       target_.target_actions(target_infoset_, history_) :
                       set<action_t> { };

  const auto r = tree.sample_sigma(infoset,
                                   targets, targeted_,
                                   average_response_,
                                   eps, gamma, rng);
  return r;
}

void oos_t::search_t::select(const tree_t& tree, rng_t &rng) {
  Expects(state_ == state_t::SELECT);
  const auto alloc = get_allocator();

  auto d = uniform_real_distribution<>();
  const auto u_delta = d(rng);
  targeted_ = (u_delta < delta_);

  if(history().player() != search_player_) {
    const auto u_eta = d(rng);
    average_response_ = (u_eta < eta_);
  }

  while (state_ == state_t::SELECT) {
    if (history_.is_terminal()) {
      prepare_suffix_probs();
      state_ = state_t::BACKPROP;
    }
    else if (history_.player() == CHANCE) {
      const auto ap = sample_chance(history_, rng, alloc);
      tree_step(ap);
    }
    else {
      const auto infoset = history_.infoset(alloc);
      const auto r = sample_tree(tree, infoset, rng);

      if (r.out_of_tree) {
        state_ = state_t::CREATE;
      }
      else {
        tree_step(r.ap, infoset);
      }
    }
  }

  Ensures(state_ == state_t::CREATE || state_ == state_t::BACKPROP);
}

void oos_t::search_t::create(tree_t& tree, rng_t &rng) {
  Expects(history_.player() != CHANCE);
  Expects(!history_.is_terminal());

  const auto infoset = history_.infoset();
  auto node = node_t(infoset.actions());

  insert_node_step(tree, infoset, node, rng);
}

void oos_t::search_t::create_prior(tree_t &tree,
                                   node_t::avg_map_t average_strategy,
                                   rng_t &rng)
{
  Expects(history_.player() != CHANCE);
  Expects(!history_.is_terminal());

  const auto infoset = history_.infoset();

  auto node = node_t(infoset.actions());

  auto sum = accumulate(begin(average_strategy),
                        end(average_strategy), (prob_t) 0.0,
                        [](const auto &r, const auto &x) {
                          return r + x.second;
                        });
  Expects(0 < sum);

  node.prior_ = move(average_strategy);

  for_each(begin(node.prior_), end(node.prior_),
           [&](auto &x) { x.second /= sum; });

  insert_node_step(tree, infoset, node, rng);
}

void oos_t::search_t::insert_node_step(tree_t &tree,
                                       const infoset_t &infoset,
                                       const node_t &node,
                                       rng_t &rng)
{
  Expects(state_ == state_t::CREATE);

  auto &nodes = tree.nodes();
  nodes.emplace(infoset, node);

  const auto r = sample_tree(tree, infoset, rng);
  Expects(!r.out_of_tree);

  tree_step(r.ap, infoset);

  if (history_.is_terminal()) {
    prepare_suffix_probs();
    state_ = state_t::BACKPROP;
  }
  else {
    state_ = state_t::PLAYOUT;
  }

  Ensures(state_ == state_t::PLAYOUT || state_ == state_t::BACKPROP);
}

void oos_t::search_t::playout_step(action_prob_t ap) {
  Expects(state_ == state_t::PLAYOUT);

  history_.act(ap.a);
  suffix_prob_.x *= ap.pr_a;

  if (history_.is_terminal()) {
    prepare_suffix_probs();
    state_ = state_t::BACKPROP;
  }

  Ensures(state_ == state_t::PLAYOUT || state_ == state_t::BACKPROP);
}

void oos_t::search_t::backprop(tree_t& tree) {
  Expects(state_ == state_t::BACKPROP);
  Expects(history_.is_terminal());

  prob_t c;
  prob_t x = suffix_prob_.x;

  const prob_t  l = suffix_prob_.l;
  const value_t u = suffix_prob_.u;

  for (auto i = rbegin(path_); i != rend(path_); ++i) {
    const auto& path_item = *i;
    const auto& acting_player = path_item.player;
    const auto& infoset = path_item.infoset;

    const auto a = path_item.action_prob.a;
    const auto pr_a = path_item.action_prob.pr_a;

    const auto pi_o = path_item.prefix_prob.pi_o;
    const auto s1 = path_item.prefix_prob.s1;
    const auto s2 = path_item.prefix_prob.s2;

    c = x;
    x = pr_a * x;

    if (acting_player == CHANCE) { // TODO make this more elegant
      continue;
    }

    auto &node = tree.lookup(infoset);

    if (acting_player == search_player_) {
      const value_t w = u * pi_o / l;
      for (const auto& a_prime : infoset.actions()) {
        value_t r;
        if (a_prime == a) {
          r = (c - x) * w;
        }
        else {
          r = -x * w;
        }

        node.regret(a_prime) += r;
      }

      node.regret_n() += 1;
    }
    else {
      const auto sigma = node.sigma_regret_matching();

      const prob_t q = delta_*s1 + (1 - delta_)*s2;
      for (const auto& a_prime : infoset.actions()) {
        const prob_t s = (pi_o * sigma.pr(infoset, a_prime)) / q;
        node.average_strategy(a_prime) += s;
      }
    }
  }

  state_ = state_t::FINISHED;
}

prob_t oos_t::search_t::targeting_ratio() const {
  auto r = prefix_prob_.s2 / prefix_prob_.s1;

  Ensures(r > 0);
  Ensures(is_normal(r));
  return r;
};

void oos_t::search_t::set_initial_weight(prob_t w) {
  Expects(0 < w);
  Expects(is_normal(w));

  prefix_prob_.s1 = w;
  prefix_prob_.s2 = w;
}

auto oos_t::search_t::get_allocator() const -> allocator_type {
  return path_.get_allocator().resource();
}

static auto sample_chance(const history_t &history, rng_t& rng,
                          game_t::action_prob_allocator_t alloc)
  -> action_prob_t
{
  const auto actions_pr = history.chance_actions(alloc);
  Expects(!actions_pr.empty());

  auto actions = action_vector { };
  auto probs = prob_vector { };

  transform(begin(actions_pr), end(actions_pr), back_inserter(actions),
            [](const pair<action_t, prob_t> &x) -> action_t {
              return x.first;
            });

  transform(begin(actions_pr), end(actions_pr), back_inserter(probs),
            [](const pair<action_t, prob_t> &x) -> prob_t {
              return x.second;
            });

  auto a_dist = discrete_distribution<>(begin(probs), end(probs));
  auto i = a_dist(rng);

  auto a = actions[i];
  auto pr_a = probs[i];
  auto rho1 = pr_a;
  auto rho2 = pr_a;
  Ensures(0 <= pr_a && pr_a <= 1);

  return { a, pr_a, rho1, rho2 };
}

static auto sample_chance(const history_t &history, rng_t& rng)
  -> action_prob_t {
  return sample_chance(history, rng, { });
}

auto history_t::sample_chance(oz::rng_t &rng) const -> action_prob_t {
  return oz::sample_chance(*this, rng);
}

static auto sample_uniform(const history_t &history, rng_t &rng)
  -> action_prob_t
{
  auto actions = history.infoset().actions();
  auto N = static_cast<int>(actions.size());
  Expects(N > 0);

  auto d = uniform_int_distribution<>(0, N-1);
  auto i = d(rng);
  auto a = actions[i];
  auto pr_a = (prob_t) 1/N;
  Ensures(0 <= pr_a && pr_a <= 1);

  return { a, pr_a, pr_a, pr_a };
}

static inline auto sample_action(const history_t &h, rng_t &rng)
  -> action_prob_t
{
  if (h.player() == CHANCE) {
    return sample_chance(h, rng);
  }
  else {
    return sample_uniform(h, rng);
  }
}

void oos_t::search_iter(history_t h, player_t player,
                        tree_t &tree, rng_t &rng,
                        target_t target,
                        infoset_t target_infoset,
                        void *buffer, size_t buffer_size,
                        const prob_t eps,
                        const prob_t delta,
                        const prob_t gamma,
                        const prob_t beta)
{
  using state_t = search_t::state_t;

  monotonic_buffer_resource buf_rsrc(buffer, buffer_size);

  search_t s(move(h), player,
             move(target), move(target_infoset),
             &buf_rsrc,
             eps, delta, gamma);

  s.set_initial_weight(1.0/avg_targeting_ratio_);

  while (s.state() != state_t::FINISHED) {
    switch (s.state()) {
      case state_t::SELECT:
        s.select(tree, rng);
        break;
      case state_t::CREATE:
        s.create(tree, rng);
        break;
      case state_t::PLAYOUT:
        s.playout_step(sample_action(s.history(), rng));
        break;
      case state_t::BACKPROP:
        s.backprop(tree);
        break;
      case state_t::FINISHED:
        break;
    }
  }

  avg_targeting_ratio_ =
      beta*avg_targeting_ratio_ + (1-beta)*s.targeting_ratio();
}

static constexpr int WORK_BUFFER_SIZE = (2 << 20);

static unique_ptr<uint8_t[]> make_work_buffer(size_t size) {
  unique_ptr<uint8_t[]> ptr(new uint8_t[size]);
  return ptr;
}

void oos_t::search_targeted(history_t h, int n_iter, tree_t &tree, rng_t &rng,
                   target_t target, infoset_t target_infoset,
                   const prob_t eps,
                   const prob_t delta,
                   const prob_t gamma,
                   const prob_t beta)
{
  Expects(n_iter >= 0);

  auto ptr = make_work_buffer(WORK_BUFFER_SIZE);

  for(int i = 0; i < n_iter; i++) {
    search_iter(h, P1, tree, rng,
                target, target_infoset,
                ptr.get(), WORK_BUFFER_SIZE,
                eps, delta, gamma, beta);

    search_iter(h, P2, tree, rng,
                target, target_infoset,
                ptr.get(), WORK_BUFFER_SIZE,
                eps, delta, gamma, beta);
  }
}

void oos_t::search(history_t h, int n_iter, tree_t &tree, rng_t &rng,
                   const prob_t eps,
                   const prob_t delta,
                   const prob_t gamma,
                   const prob_t beta)
{
  search_targeted(move(h), n_iter, tree, rng,
                  null_target(), null_infoset(),
                  eps, delta, gamma, beta);
}

} // namespace oz

File: oz/oos.py
from copy import copy
from collections import defaultdict

import numpy.random
import numpy as np

def sample(h, context, infoset, sigma, s1, s2, i):
    # FIXME implement targeting
    # a, rho1, rho2 = sigma.sample_targeted(context, infoset, eps)
    # return a, rho1*s1, rho2*s2
    player = h.player
    if player == i:
        eps = context.eps
        a, pr_a = sigma.sample_eps(infoset, eps)
        return a, pr_a*s1, pr_a*s2
    else:
        a, pr_a = sigma.sample_pr(infoset)
        return a, pr_a*s1, pr_a*s2


def sample_chance(h, context):
    rng = context.rng
    infoset = h.infoset()
    actions = infoset.actions
    probs = infoset.probs
    i = rng.choice(len(actions), p=probs)
    a = actions[i]
    pr_a = probs[i]
    return a, pr_a, pr_a


def playout(h, s, sigma):
    # FIXME chance player
    x = 1.
    while not h.is_terminal():
        infoset = h.infoset()
        a, pr_a = sigma.sample_pr(infoset)
        h = h >> a
        x = pr_a*x
    u = h.utility()
    return x, s*x, u


def _regret_to_pr(regrets):
    total = sum(v for v in regrets.values() if v > 0)

    probs = {}
    if total > 0:
        for a, r in regrets.items():
            if r > 0:
                probs[a] = r / total
            else:
                probs[a] = 0
    else:
        p = 1./len(regrets)
        for a in regrets:
            probs[a] = p

    return probs


class SigmaUniform:
    def __init__(self, rng):
        self._rng = rng

    def pr(self, infoset, a):
        actions = infoset.actions
        return 1./len(actions)

    def sample_pr(self, infoset):
        rng = self._rng
        actions = infoset.actions
        a = rng.choice(actions)
        pr_a = 1./len(actions)
        return a, pr_a

    def sample_eps(self, infoset, eps):
        return self.sample_pr(infoset)


class SigmaRegretMatching:
    def __init__(self, regrets, rng):
        self.regrets = regrets
        self._rng = rng
        pr = _regret_to_pr(regrets)
        actions = list(pr.keys())
        actions.sort(key=lambda x: x.name)
        probs = [pr[a] for a in actions]
        self._pr = pr
        self._actions = actions
        self._probs = probs

    def pr(self, infoset, a):
        return self._pr[a]

    def sample_pr(self, infoset):
        rng = self._rng
        a = rng.choice(self._actions, p=self._probs)
        return a, self._pr[a]

    def sample_eps(self, infoset, eps):
        rng = self._rng
        if rng.rand() > eps:
            a = rng.choice(self._actions, p=self._probs)
        else:
            a = rng.choice(self._actions)
        pr_a_eps = eps*(1./len(self._actions)) + (1-eps)*self._pr[a]
        return a, pr_a_eps


class SigmaAverageStrategy:
    def __init__(self, tree, rng):
        self.tree = tree
        self._rng = rng

    def pr(self, infoset, a, uniform_fallback=True):
        rng = self._rng
        if uniform_fallback:
            node = self.tree.nodes.get(infoset)
        else:
            node = self.tree.nodes[infoset]

        if node is not None:
            total = sum(node.average_strategy.values())
            if total == 0.0:
                actions = infoset.actions
                return 1. / len(actions)
            return float(node.average_strategy[a]) / total
        else:
            actions = infoset.actions
            return 1. / len(actions)

    def sample_pr(self, infoset, uniform_fallback=True):
        rng = self._rng
        if uniform_fallback:
            node = self.tree.nodes.get(infoset)
        else:
            node = self.tree.nodes[infoset]

        if node is not None:
            average_strategy = node.average_strategy
            actions = average_strategy.keys()
            values = average_strategy.values()
            total = sum(values)
            probs = [v / total for v in values]
            a = rng.choice(actions, p=probs)
            pr_a = average_strategy[a] / total
            return a, pr_a
        else:
            actions = infoset.actions
            a = rng.choice(actions)
            pr_a = 1. / len(actions)
            return a, pr_a

    def sample_eps(self, infoset, eps):
        rng = self._rng
        node = self.tree.nodes[infoset]
        average_strategy = node.average_strategy
        actions = average_strategy.keys()
        values = average_strategy.values()
        total = sum(values)
        if rng.rand() > eps:
            probs = [v / total for v in values]
            a = rng.choice(actions, p=probs)
        else:
            a = rng.choice(actions)
        pr_a = average_strategy[a] / total
        pr_a_eps = eps*(1./len(actions)) + (1-eps)*pr_a
        return a, pr_a_eps


class Tree:
    class Node:
        def __init__(self):
            self.regrets = defaultdict(float)
            self.average_strategy = defaultdict(float)

        def sigma_regret_matching(self, rng):
            # TODO Make this less ugly
            if len(self.regrets) > 0:
                return SigmaRegretMatching(self.regrets, rng=rng)
            else:
                return SigmaUniform(rng=rng)

        def update_regret(self, a, r):
            self.regrets[a] += r

        def update_average_strategy(self, a, s):
            self.average_strategy[a] += s

    def __init__(self):
        self.nodes = {}

    def lookup(self, infoset):
        if infoset in self.nodes:
            return self.nodes[infoset], False
        else:
            node = self.Node()
            self.nodes[infoset] = node
            return node, True

    def sigma_average_strategy(self, rng):
        return SigmaAverageStrategy(self, rng)


class Context:
    def __init__(self, eps=0.5, sigma=0.4, seed=None):
        self.rng = numpy.random.RandomState(seed=seed)
        self.sigma_playout = SigmaUniform(rng=self.rng)
        self.eps = eps
        self.delta = sigma


def oos(h, context, tree, pi_i, pi_o, s1, s2, i):
    delta = context.delta
    rng = context.rng
    player = h.player

    if h.is_terminal():
        l = delta*s1 + (1.-delta)*s2
        u = h.utility(i)
        return 1, l, u

    elif player == h.Player.Chance:
        (a, rho1, rho2) = sample_chance(h, context)
        (x, l, u) = oos(h >> a, context, tree,
                        pi_i, rho2 * pi_o,
                        rho1 * s1, rho2 * s2, i)
        return rho2*x, l, u

    infoset = h.infoset()
    (node, out_of_tree) = tree.lookup(infoset)

    if out_of_tree:
        sigma = context.sigma_playout
    else:
        sigma = node.sigma_regret_matching(rng)

    (a, s1_prime, s2_prime) = sample(h, context, infoset, sigma,
                                     s1, s2, i)

    pr_a = sigma.pr(infoset, a)

    if out_of_tree:
        q = delta*s1 + (1.-delta)*s2
        (x, l, u) = playout(h >> a, pr_a*q, sigma)
    else:
        if player == i:
            pi_prime_i = pr_a*pi_i
            pi_prime_o = pi_o
        else:
            pi_prime_i = pi_i
            pi_prime_o = pr_a*pi_o

        (x, l, u) = oos(h >> a, context, tree,
                        pi_prime_i, pi_prime_o,
                        s1_prime, s2_prime, i)

    c = x
    x = pr_a*x

    if player == i:
        w = u*pi_o / l
        for a_prime in infoset.actions:
            if a_prime == a:
                r = (c - x)*w
            else:
                r = - x*w
            node.update_regret(a_prime, r)

    else:
        q = delta*s1 + (1.-delta)*s2
        for a_prime in infoset.actions:
            s = (1./q) * pi_o * sigma.pr(infoset, a_prime)
            node.update_average_strategy(a_prime, s)

    return x, l, u


def solve(h, context, tree, n_iter):
    for i in range(n_iter):
        oos(copy(h), context, tree, 1, 1, 1, 1, h.Player.P1)
        oos(copy(h), context, tree, 1, 1, 1, 1, h.Player.P2)
